### PR TITLE
Shell escape

### PR DESCRIPTION
**Pull request by nhazekam**
@btovar Take a look. Slightly confusing, but there was an older escape_shell_string that I am told is used for hadoop. Naming changes and whatnot can be addressed as needed just wanted to get this to you.

### CODE REVIEW

**Comment by dthain** [comment]
@nhazekam it would be helpful if you can look up some reference for `/bin/sh` so that you can be sure you have identified everything that needs to be escaped.

**Comment by nhazekam** [comment]
As per @batrick 's suggestion I looked at the opengroup documentation of the Shell Command Langauge.

In section 2.2.3 on Double-Quoted Strings, it indicates you only need to escape dollar sign, backtick, and backslash. I also escape double quote as we are adding and exterior double quote around the string.

http://pubs.opengroup.org/onlinepubs/009695399/utilities/xcu_chap02.html#tag_02

I can look at other resources/add more if I missed any.

File: dttools/src/stringtools.c [hunk 1]
```diff
@@ -46,6 +47,27 @@ char *escape_shell_string(const char *str)
 	return escaped_string;
 }
 
+char *string_escape_shell( const char *str )
+{
+	buffer_t buffer;
```
Comment (batrick) [inline dttools/src/stringtools.c]: I find it helpful to use an array here so you don't need to use `&` everywhere:

`buffer_t B[1];`

then you can do:

`buffer_putliteral(B, "foo");`

File: dttools/src/stringtools.c [hunk 2]
```diff
@@ -46,6 +47,27 @@ char *escape_shell_string(const char *str)
...
+	buffer_init(&buffer);
+
+	char *s;
+	buffer_putlstring(&buffer,"\"",1);
```
Comment (batrick) [inline dttools/src/stringtools.c]: Use `buffer_putliteral(&buffer, "\"");`.

**Comment by dthain** [comment]
That's great, pls add that reference to the code..

File: dttools/src/stringtools.c [hunk 3]
```diff
@@ -46,6 +47,27 @@ char *escape_shell_string(const char *str)
 	return escaped_string;
 }
 
+char *string_escape_shell( const char *str )
+{
+	buffer_t buffer;
+	buffer_init(&buffer);
```
Comment (batrick) [inline dttools/src/stringtools.c]: I'm guessing you also want `bufer_abortonfailure` since you're not checking for (memory) errors when using the buffer.

**Comment by nhazekam** [comment]
As per @batrick 's suggestion I looked at the opengroup documentation of
the Shell Command Langauge.

In section 2.2.3 on Double-Quoted Strings, it indicates you only need to
escape dollar sign, backtick, and backslash. I also escape double quote as
we are adding and exterior double quote around the string.

http://pubs.opengroup.org/onlinepubs/009695399/utilities/xcu_chap02.html#tag_02

File: dttools/src/stringtools.c [hunk 4]
```diff
@@ -839,14 +861,24 @@ char * string_wrap_command( const char *command, const char *wrapper_command )
 {
 	if(!wrapper_command) return strdup(command);
 
-	char * result = malloc(strlen(command)+strlen(wrapper_command)+2);
 	char * braces = strstr(wrapper_command,"{}");
+	char * square = strstr(wrapper_command,"[]");
+
+	if(square)
+		command = string_escape_shell(command);
```
Comment (batrick) [inline dttools/src/stringtools.c]: forgot `free(command)`

**Comment by dthain** [comment]
Don't forget to add something to the manual.

**Comment by btovar** [comment]
Superseded by #956

**Pull request closed without merging**